### PR TITLE
refactor: move discovery and util under package app/customresource to make packages highly cohesive

### LINKED ISSUE #2103: Potential cyclic imports
**What happened**:
Encounter some cyclic imports errors after rebasing a WIP PR to master. 
It means the code structure is lowly cohesive/ highly coupling, or the WIP PR needs refactoring imports.

Compared dependencies and found these newly added dependencies:
```
v2/pkg/app
├──v2/pkg/util
├──v2/internal/discovery

v2/pkg/customresourcestate
├──v2/pkg/util
├──v2/internal/discovery

v2/internal/store
├──v2/pkg/util

v2/pkg/util
└──v2/pkg/customresource
v2/internal/discovery
├──v2/pkg/metricshandler
├──v2/pkg/util
├──v2/internal/store
├──v2/pkg/customresource
└──v2/pkg/options
```
```
# before
godepgraph ./

v2/pkg/app
├──v2/pkg/metricshandler
├──v2/pkg/allowdenylist
├──v2/pkg/customresource
├──v2/pkg/customresourcestate
├──v2/pkg/options
├──v2/internal/store
├──v2/pkg/optin
├──v2/pkg/util/proc
└──v2/pkg/metric_generator


v2/pkg/customresourcestate
├──v2/pkg/metric_generator
├──v2/pkg/customresource
└──v2/pkg/metric



v2/pkg/optin
└──v2/pkg/metric_generator
v2/pkg/sharding
v2/pkg/util/proc
v2/pkg/allowdenylist
└──v2/pkg/metric_generator
v2/pkg/builder/types
├──v2/pkg/metrics_store
├──v2/pkg/customresource
├──v2/pkg/metric_generator
└──v2/pkg/options
v2/pkg/customresource
└──v2/pkg/metric_generator
v2/pkg/watch
v2/tests/e2e/framework
v2/tools
v2/internal
├──v2/pkg/app
└──v2/pkg/options
v2/internal/store
├──v2/pkg/watch
├──v2/pkg/customresource
├──v2/pkg/metric_generator
├──v2/pkg/options
├──v2/pkg/metrics_store
├──v2/pkg/metric
├──v2/pkg/sharding
├──v2/pkg/constant
└──v2/pkg/builder/types

v2/pkg/allow
v2/pkg/options
v2
├──v2/internal
└──v2/pkg/options
v2/pkg/builder
├──v2/internal/store
├──v2/pkg/builder/types
├──v2/pkg/customresource
├──v2/pkg/metric_generator
├──v2/pkg/metrics_store
└──v2/pkg/options
v2/pkg/metric
v2/pkg/metric_generator
└──v2/pkg/metric
v2/pkg/metrics_store/metricsstore
└──v2/pkg/metric
v2/pkg/metricshandler
├──v2/pkg/builder/types
├──v2/pkg/metrics_store
└──v2/pkg/options
```

```
# now
godepgraph ./

v2/pkg/app
├──v2/pkg/util
├──v2/internal/discovery
├──v2/pkg/optin
├──v2/pkg/customresourcestate
├──v2/pkg/metricshandler
├──v2/pkg/options
├──v2/pkg/allowdenylist
├──v2/internal/store
├──v2/pkg/metric_generator
├──v2/pkg/util/proc
└──v2/pkg/customresource
v2/pkg/customresourcestate
├──v2/pkg/util
├──v2/internal/discovery
└──v2/internal/store
├──v2/pkg/customresource
├──v2/pkg/metric
├──v2/pkg/metric_generator
v2/pkg/optin
└──v2/pkg/metric_generator
v2/pkg/sharding
v2/pkg/util/proc
v2/pkg/allowdenylist
└──v2/pkg/metric_generator
v2/pkg/builder/types
├──v2/pkg/options
├──v2/pkg/metrics_store
├──v2/pkg/customresource
└──v2/pkg/metric_generator
v2/pkg/customresource
└──v2/pkg/metric_generator
v2/pkg/watch
v2/tests/e2e/framework
v2/tools
v2/internal
├──v2/pkg/app
└──v2/pkg/options
v2/internal/store
├──v2/pkg/util
├──v2/pkg/sharding
├──v2/pkg/builder/types
├──v2/pkg/options
├──v2/pkg/metric
├──v2/pkg/metric_generator
├──v2/pkg/constant
├──v2/pkg/metrics_store
├──v2/pkg/watch
└──v2/pkg/customresource
v2/pkg/allow
v2/pkg/options
v2
├──v2/internal
└──v2/pkg/options
v2/pkg/builder
├──v2/internal/store
├──v2/pkg/builder/types
├──v2/pkg/metric_generator
├──v2/pkg/options
├──v2/pkg/customresource
└──v2/pkg/metrics_store
v2/pkg/metric
v2/pkg/metric_generator
└──v2/pkg/metric
v2/pkg/metrics_store/metricsstore
└──v2/pkg/metric
v2/pkg/metricshandler
├──v2/pkg/builder/types
├──v2/pkg/options
└──v2/pkg/metrics_store

v2/pkg/util
└──v2/pkg/customresource
v2/internal/discovery
├──v2/pkg/metricshandler
├──v2/pkg/util
├──v2/internal/store
├──v2/pkg/customresource
└──v2/pkg/options
```
**What you expected to happen**:
Refactor some codes to make sure it's  highly cohesive  and lowly coupling.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- kube-state-metrics version: 
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Other info:

### PR DESCRIPTION
**Pull request by CatherineF-dev**
**What this PR does / why we need it**:

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2103

### CODE REVIEW

**Comment by CatherineF-dev** [comment]
Ready for reviewing~

File: pkg/app/server.go [hunk 1]
```diff
@@ -51,7 +54,8 @@ import (
 	"k8s.io/kube-state-metrics/v2/pkg/metricshandler"
 	"k8s.io/kube-state-metrics/v2/pkg/optin"
 	"k8s.io/kube-state-metrics/v2/pkg/options"
-	"k8s.io/kube-state-metrics/v2/pkg/util"
+
+	// "k8s.io/kube-state-metrics/v2/pkg/util"
```
Comment (logicalhan) [inline pkg/app/server.go:58]: ?

File: pkg/app/server.go [hunk 2]
```diff
@@ -453,3 +458,74 @@ func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.Con
...
+	// figure out if apiserver is configured incorrectly.
+	klog.InfoS("Tested communication with server")
+	v, err := kubeClient.Discovery().ServerVersion()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error while trying to communicate with apiserver: %w", err)
+	}
+	klog.InfoS("Run with Kubernetes cluster version", "major", v.Major, "minor", v.Minor, "gitVersion", v.GitVersion, "gitTreeState", v.GitTreeState, "gitCommit", v.GitCommit, "platform", v.Platform)
+	klog.InfoS("Communication with server successful")
+
+	return kubeClient, customResourceClients, nil */
```
Comment (logicalhan) [inline pkg/app/server.go:497]: Why is this commented out?

File: pkg/customresource/utils.go [hunk 3]
```diff
@@ -14,73 +14,28 @@ See the License for the specific language governing permissions and
...
-	"k8s.io/kube-state-metrics/v2/pkg/customresource"
 )
 
 var config *rest.Config
 var currentKubeClient clientset.Interface
 var currentDiscoveryClient *discovery.DiscoveryClient
 var currentDynamicClient *dynamic.DynamicClient
 
-// CreateKubeClient creates a Kubernetes clientset and a custom resource clientset.
-func CreateKubeClient(apiserver string, kubeconfig string) (clientset.Interface, error) {
```
Comment (rexagod) [inline pkg/customresource/utils.go:44]: I'm not sure why was this moved out of here?

Comment (rexagod) [inline pkg/customresource/utils.go:44]: I'm not sure why this was moved out?

File: pkg/app/server.go (hunk 1 above)
Comment (CatherineF-dev) [reply pkg/app/server.go]: Done. Cleaned up.

File: pkg/app/server.go (hunk 2 above)
Comment (CatherineF-dev) [reply pkg/app/server.go]: Done. Cleaned up.

File: pkg/customresource/utils.go (hunk 3 above)
Comment (CatherineF-dev) [reply pkg/customresource/utils.go:44]: Remove this dependency
v2/pkg/util
└──v2/pkg/customresource

https://github.com/kubernetes/kube-state-metrics/issues/2103

**Comment by CatherineF-dev** [comment]
/hold

I find it still has these newly added dependencies.  Will try to remove these as well.
```
v2/pkg/customresourcestate
├──v2/pkg/metricshandler
├──v2/pkg/options
├──v2/internal/store
```

File: pkg/app/discovery.go [hunk 4]
```diff
@@ -12,13 +12,15 @@ limitations under the License.
 */
 
 // Package discovery provides a discovery and resolution logic for GVKs.
-package discovery
+package app
```
[the remark this answers is not in the document]
Comment (CatherineF-dev) [reply pkg/app/discovery.go:14]: Done.

**Comment by CatherineF-dev** [comment]
/unhold

Current godepgraph is 
```
v2/pkg/builder/types
├──v2/pkg/metric_generator
├──v2/pkg/options
├──v2/pkg/metrics_store
└──v2/pkg/customresource
v2/pkg/metric_generator
└──v2/pkg/metric
v2/pkg/options
v2/tests/e2e/framework
v2/pkg/builder
├──v2/pkg/builder/types
├──v2/pkg/metric_generator
├──v2/internal/store
├──v2/pkg/customresource
├──v2/pkg/metrics_store
└──v2/pkg/options
v2/pkg/customresource
└──v2/pkg/metric_generator
v2/pkg/customresourcestate
├──v2/pkg/metric
├──v2/pkg/metric_generator
├──v2/internal/discovery
├──v2/pkg/customresource
└──v2/internal/store
v2/pkg/watch
v2/pkg/sharding
v2/pkg/util/proc
v2/internal
├──v2/pkg/options
└──v2/pkg/app
v2/internal/store
├──v2/pkg/metrics_store
├──v2/pkg/sharding
├──v2/pkg/watch
├──v2/pkg/metric_generator
├──v2/pkg/builder/types
├──v2/pkg/constant
├──v2/pkg/customresource
├──v2/pkg/metric
└──v2/pkg/options
v2/pkg/allow
v2/pkg/allowdenylist
└──v2/pkg/metric_generator
v2/pkg/metricshandler
├──v2/pkg/options
├──v2/pkg/metrics_store
└──v2/pkg/builder/types
v2/pkg/optin
└──v2/pkg/metric_generator
v2
├──v2/internal
└──v2/pkg/options
v2/internal/discovery
v2/pkg/app
├──v2/pkg/optin
├──v2/pkg/customresource
├──v2/pkg/allowdenylist
├──v2/pkg/util/proc
├──v2/pkg/customresourcestate
├──v2/pkg/options
├──v2/pkg/metric_generator
├──v2/internal/store
├──v2/internal/discovery
└──v2/pkg/metricshandler
v2/pkg/metric
v2/pkg/metrics_store/metricsstore
└──v2/pkg/metric
v2/tools
```

**Comment by CatherineF-dev** [comment]
Ping ~ @rexagod @mrueg

**Comment by rexagod** [comment]
Sorry for the delay here, I got caught up, will review before end of the day. 🙂

File: pkg/app/discovery.go [hunk 5]
```diff
@@ -27,16 +28,87 @@ import (
...
 	"k8s.io/kube-state-metrics/v2/pkg/customresource"
 	"k8s.io/kube-state-metrics/v2/pkg/metricshandler"
 	"k8s.io/kube-state-metrics/v2/pkg/options"
-	"k8s.io/kube-state-metrics/v2/pkg/util"
 )
 
 // Interval is the time interval between two cache sync checks.
 const Interval = 3 * time.Second
 
+// CRDiscoverer provides a cache of the collected GVKs, along with helper utilities.
```
Comment (rexagod) [inline pkg/app/discovery.go:41]:
```suggestion
// CRDiscoverer provides a cache of the collected GVKs, along with helper utilities.
// It implements customresourcestate.GVKToGVKP.
```

File: pkg/customresourcestate/config.go [hunk 6]
```diff
@@ -166,8 +165,13 @@ type ConfigDecoder interface {
 	Decode(v interface{}) (err error)
 }
 
+// GVKToGVKP interface implements ResolveGVKToGVKPs
```
Comment (rexagod) [inline pkg/customresourcestate/config.go:168]:
```suggestion
```

File: internal/discovery/types.go [hunk 7]
```diff
@@ -15,93 +15,22 @@ package discovery
...
 import (
 	"fmt"
-	"sync"
 
-	"github.com/prometheus/client_golang/prometheus"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-type groupVersionKindPlural struct {
+// GroupVersionKindPlural is GVK + Plural
```
Comment (rexagod) [inline internal/discovery/types.go:22]:
```suggestion
// GroupVersionKindPlural embeds schema.GroupVersionKind, in addition to the plural representation of the Kind.
```
Nit.

File: internal/discovery/types.go [hunk 8]
```diff
@@ -15,93 +15,22 @@ package discovery
...
 	Plural string
 }
 
-func (g groupVersionKindPlural) String() string {
+func (g GroupVersionKindPlural) String() string {
 	return fmt.Sprintf("%s/%s, Kind=%s, Plural=%s", g.Group, g.Version, g.Kind, g.Plural)
 }
 
-type kindPlural struct {
+// KindPlural is Kind + Plural
```
Comment (rexagod) [inline internal/discovery/types.go:32]:
```suggestion
// KindPlural entails the Kind and its plural representation.
```
Nit.

File: pkg/app/server.go [hunk 9]
```diff
@@ -453,3 +457,38 @@ func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.Con
 	}
 	return nil, nil
 }
+
+func createKubeClient(apiserver string, kubeconfig string) (clientset.Interface, error) {
```
[the remark this answers is not in the document]
Comment (rexagod) [reply pkg/app/server.go:461]: This would mean that only the logic that's truly incorporated repository-wide ends up in the higher-hierarchy `utils` package.

**Review by rexagod — CHANGES_REQUESTED** [review]

File: internal/discovery/types.go [hunk 10]
```diff
@@ -15,93 +15,22 @@ package discovery
```
[the remark this answers is not in the document]
Comment (CatherineF-dev) [reply internal/discovery/types.go:15]: If moving entire discovery package, v2/pkg/customresourcestate will depends on all packages where internal/discovery depends on. 

https://github.com/kubernetes/kube-state-metrics/issues/2103

[the remark this answers is not in the document]
Comment (rexagod) [reply internal/discovery/types.go:15]: ACK, can we move the `discovery` types inside `pkg/app` to it's own package (could be `types`, which would contain all such exported types for the `app` package, since `discovery` comes under that now)?

File: pkg/customresourcestate/config.go [hunk 11]
```diff
@@ -166,8 +165,12 @@ type ConfigDecoder interface {
 	Decode(v interface{}) (err error)
 }
 
+type GVKToGVKP interface {
```
Comment (rexagod) [inline pkg/customresourcestate/config.go:168]:
```suggestion
// GVKToGVKP incorporates logic to resolve a given GVK, with variable version and/or kind, to it's []discovery.GroupVersionKindPlural representation.
type GVKToGVKP interface {
```

Looks like we need to annotate this, so the linter is happy.

File: internal/discovery/types.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (CatherineF-dev) [reply internal/discovery/types.go:15]: discovery.go and internal/discovery/types.go should be in different packages. Because customresourcestate depends on internal/discovery/types.go and discovery.go depends on many other pkgs.

Does `can we move the discovery types inside pkg/app to it's own package` mean moving internal/discovery/types.go under `types` and discovery.go under `app`?

[the remark this answers is not in the document]
Comment (rexagod) [reply internal/discovery/types.go:15]: Yep, although on second thought, it seems more sensible to move such types to a `types` package, which, given the fact that its used in `internal` and `pkg`, should ideally live at the project root. We can add other types that are used repository-wide in that package itself, in the future.

@mrueg WDYT?

[the remark this answers is not in the document]
Comment (rexagod) [reply internal/discovery/types.go:15]: My bad, it seems the `discovery` types are not used in `internal`, so moving these types within a `types` package in `pkg` is a valid alternative, in case we do not want to have a root-level `types` package.

[the remark this answers is not in the document]
Comment (mrueg) [reply internal/discovery/types.go:15]: I'm not sure if it's worth reorganizing when we want to change the way customresource metrics are created at the same time, so I don't have a strong opinion here.

[the remark this answers is not in the document]
Comment (rexagod) [reply internal/discovery/types.go:15]: ACK. In that case I think we can proceed with a `types` package within `pkg` that encapsulates all types used throughout `pkg`, instead of a root-level `types` package. This would ensure relevant types live close to their logic, while providing a degree of desired isolation away from the package the logic is present under.

@CatherineF-dev Does this sound good?

[the remark this answers is not in the document]
Comment (CatherineF-dev) [reply internal/discovery/types.go:15]: QQ: what's the package name for types.go? pkg or types.

[the remark this answers is not in the document]
Comment (rexagod) [reply internal/discovery/types.go:15]: `types` should suffice I belive.
